common/datasets/sms_wsj/returnn_datasets.py

```diff
@@ -0,0 +1,617 @@
+        """
+        data = self[seq_idx]
+        d = {k: v.size for k, v in data.items()}
+        for update_key in ["data", "target_signals"]:
+            if update_key in d and "seq_len" in data:
+                d[update_key] = int(data["seq_len"])
+        return NumbersDict(d)
+
+    def update_buffer(self, seq_idx: int, pop_seqs: bool = True):
+        """
+        :param int seq_idx:
+        :param bool pop_seqs: if True, pop sequences from buffer that are outside buffer range
+        """
+        if not self._use_buffer:
+            return
+
+        # debugging information
+        keys = list(self._buffer.keys()) or [0]
+        if not (min(keys) <= seq_idx <= max(keys)):
+            print(
+                f"WARNING: seq_idx {seq_idx} outside range of keys: {self._buffer.keys()}",
+                file=returnn_log.v5,
+            )
+
+        # add sequences
+        for idx in range(seq_idx, min(seq_idx + self._buffer.max_size // 2, len(self))):
+            if idx not in self._buffer:
+                self._buffer[idx] = next(self._ds_iterator)
+            if idx == len(self) - 1 and 0 not in self._buffer:
+                print(f"Reached end of dataset, reset iterator", file=returnn_log.v4)
+                try:
+                    next(self._ds_iterator)
+                except StopIteration:
+                    pass
+                else:
+                    print(
+                        "WARNING: reached final index of dataset, but iterator has more sequences. "
+                        "Maybe the training was restarted from an epoch > 1?",
+                        file=returnn_log.v3,
+                    )
+                print(
+                    f"Current buffer indices: {self._buffer.keys()}",
+                    file=returnn_log.v5,
+                )
+                self._ds_iterator = iter(self._ds)
+                for idx_ in range(min(self._buffer.max_size // 2, len(self))):
+                    if idx_ not in self._buffer:
+                        self._buffer[idx_] = next(self._ds_iterator)
+                print(
+                    f"After adding start of dataset to buffer indices: {self._buffer.keys()}",
+                    file=returnn_log.v5,
+                )
+
+    @staticmethod
+    def _cache_zipped_audio(zip_cache: str, json_path: str, dataset_name: str):
+        """
+        Caches and unzips a given archive with SMS-WSJ data which will then be used as data dir.
+        This is done because caching of the single files takes extremely long.
+        """
+        print(f"Cache and unzip SMS-WSJ data from {zip_cache}", file=returnn_log.v4)
+
+        # cache file
+        try:
+            zip_cache_cached = sp.check_output(["cf", zip_cache]).strip().decode("utf8")
+            assert (
+                zip_cache_cached != zip_cache
+            ), "cached and original file have the same path"
+            local_base_dir = os.path.dirname(zip_cache_cached)
```

Comment:
I think the definition of `local_base_dir` has to be changed. When trying to run a train-job, the zip-file I used was already cached in Simon's tmp-directory and consequently it also wanted to unzip it there and not in my own tmp-directory.
Maybe `local_base_dir` could be a global variable or something, to make sure your proper directory is used.

Reply:
The idea was that it is desired to reuse that unzipped folder. So if it already exists, why creating another copy of it?
`unzip -n` should already do that, but it does not work if the user doesn't have write permissions, that's why I added the `chmod -R o+w <local_base_dir>`. There might be nodes in the cluster where the folder already exists without proper permissions. However, I'm not sure if there might be more issues than this.

Reply:
I forgot to say that I had to remove `chmod -R o+w <local_base_dir>` because I had no permission to change permissions...

Reply:
I removed that logic and read the data directly from the zip file now which should make this discussion obsolete.